pypeds/shape2d.py

```diff
@@ -99,41 +156,56 @@ def distance(cls, shape, other) -> (float, Vector2D):
-                dis = other.dist(shape)
-                return dis, ((other.x - shape.x) / dis, (other.y - shape.y) / dis)
-            if isinstance(shape, Circle2D):
-                dis, dir = DistanceCalculator.distance(shape.center, other)
-                return dis - shape.radius, dir
-            if isinstance(shape, Ellipse2D):
-                dis, dir = DistanceCalculator.distance(other, shape)
-                return dis, ((-1) * dir.x, (-1) * dir.y)
-            if isinstance(shape, Box2D):
-                pass
-            if isinstance(shape, Rectangle2D):
-                pass
+                dist = other.dist(shape)
+                return dist, ((other.x - shape.x) / dist, (other.y - shape.y) / dist)
+            if isinstance(shape, Circle2D or Ellipse2D):
+                dist, dirt = DistanceCalculator.distance(other, shape)
+                return dist, dirt.mul(-1)
+            ## TODO When the shape contains centers of shape and other, dirt calculating maybe is error
+            if isinstance(shape, Box2D or Rectangle2D):
+                dist1, dirt1 = DistanceCalculator.distance(shape.get_left, other)
+                dist2, dirt2 = DistanceCalculator.distance(shape.get_right, other)
+                dist3, dirt3 = DistanceCalculator.distance(shape.get_down, other)
+                dist4, dirt4 = DistanceCalculator.distance(shape.get_up, other)
+                dist = (dist1, dist2, dist3, dist4)
+                dirt = (dirt1, dirt2, dirt3, dirt4)
+                if not shape.contains(other):
+                    return min(dist), dirt[dist.index(min(dist))]
+                return - min(dist), dirt[dist.index(min(dist))].mul(-1)
         if isinstance(other, Circle2D):
-            dis, dir = DistanceCalculator.distance(shape, other.center)
-            return dis - other.radius, dir
+            dist, dirt = DistanceCalculator.distance(shape, other.center)
+            dist -= other.radius
+            if dist <= 0:
+                return dist, dirt
+            return dist, dirt.mul(-1)
         if isinstance(other, Ellipse2D):
-            l_dis, l_dir = DistanceCalculator.distance(shape, other.c_left)
-            r_dis, r_dir = DistanceCalculator.distance(shape, other.c_right)
-            m_dis, m_dir = DistanceCalculator.distance(shape, other.center)
-            l_dis -= (other.a - other.b) / 2
-            r_dis -= (other.a - other.b) / 2
-            m_dis -= other.b
-            if l_dis < r_dis:
-                if l_dis < m_dis:
-                    return l_dis, l_dir
+            l_dist, l_dirt = DistanceCalculator.distance(shape, other.c_left)
+            r_dist, r_dirt = DistanceCalculator.distance(shape, other.c_right)
+            m_dist, m_dirt = DistanceCalculator.distance(shape, other.center)
+            l_dist -= (other.a - other.b) / 2
+            r_dist -= (other.a - other.b) / 2
+            m_dist -= other.b
+            if l_dist < r_dist:
```

Comment:
compare l_dist, m_dist and r_dist by **min()** or **max()**, no need to use **if** and **else** like Java